### PR TITLE
Cli entry point

### PR DESCRIPTION
**Pull request by bollu**
This adds an entry point to parse MLIR code and execute it via our denotational semantics.

### CODE REVIEW

**Comment by bollu** [comment]
@goens I suggest we do this in a follow-up PR, but we should test that running our entrypoint actually applies the rewrites. I've historically used the [`llvm-lit`](https://llvm.org/docs/CommandGuide/lit.html) infrastructure for this type of test.

**Comment by goens** [comment]
@bollu sure! maybe we want to open an issue to keep track of that later?

**Comment by bollu** [comment]
@goens I pushed some changes, mainly to reflow `Parser.lean`. Let me know if the changes work for you!

**Comment by bollu** [comment]
@goens I think we've now moved the `Transform` machinery into `SSA/Core/`,  so we should delete that from this PR, and include it from `SSA/Core`.

**Comment by goens** [comment]
@bollu I was working on that :( I'm not sure now,  but I would appreciate if you would coordinate before doing that next time

**Comment by bollu** [comment]
Aha, go ahead and revert @goens !

**Comment by bollu** [comment]
I was merely procrastinating from admin work 😄

**Comment by bollu** [comment]
And yes, I'll ping before I snipe some low hanging PR fruit next time (:

File: lake-manifest.json
Comment (tobiasgrosser) [inline lake-manifest.json:7]: I am curious why the std library changes in this PR?

Comment (goens) [reply lake-manifest.json:7]: That was unintentional (my lake-manifest was broken from the merge). I can revert that change.

**Comment by goens** [comment]
@bollu sorry, I did try to merge but you had moved all the whitespace around and that broke the semantic changes I was working on

File: lake-manifest.json
Comment (tobiasgrosser) [reply lake-manifest.json:7]: That would be good. 😄

**Comment by bollu** [comment]
@goens : quickly noting that I no longer remember the difference between the `opt` executable and the `mlirnatural` executable. Could you jog my memory for me please?

**Comment by tobiasgrosser** [comment]
You probably want to merge main after the mathlib bump is in to reduce the diff shown here.

**Comment by goens** [comment]
> @goens could you please edit the changes to the CI in [`.github/workflows/ci.yml`](https://github.com/opencompl/ssa/pull/102/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03f) to test the `opt` that we build? what's the correct way to run the test?

what you added there is the right way! (it's just not working yet, that's a different thing)

**Comment by tobiasgrosser** [comment]
> > You probably want to merge main after the mathlib bump is in to reduce the diff shown here.
> 
> yep! that was precisely the point of that bump :)

For some reason it looks as if you cherry-picked the relevant commits instead of doing an `git remote update; git merge origin/main`. Without the actual merge commit, github will not use the new main as base for its diff, no? Maybe you are still waiting for the merge queue to finish. In that case, just ignore this comment.

File: SSA/Core/MLIRSyntax/AST.lean
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/AST.lean:224]: Unrelated change.

**Comment by bollu** [comment]
It maybe that we need to substitute the metavars that have been assigned in the `Meta` monad context.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/Parser.lean:98]: Leftover debug statement?

File: lakefile.lean
Comment (tobiasgrosser) [inline lakefile.lean:18]: I am curious why we need the `supportInterpreter` here.

File: .github/workflows/ci.yml
Comment (tobiasgrosser) [inline .github/workflows/ci.yml:42]: whitespace? A

**Comment by tobiasgrosser** [comment]
How lovely. Thank you, @goens for pushing this through!

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:98]: yeah, I like to keep those around in case things break again :sweat_smile: don't know what the ideal way is to do that

File: lakefile.lean
Comment (goens) [reply lakefile.lean:18]: that's because we use the Lean frontend in our code (we call the Lean parser), so that it also links it.

Comment (tobiasgrosser) [reply lakefile.lean:18]: Maybe add a comment explaining this.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (tobiasgrosser) [reply SSA/Core/MLIRSyntax/Parser.lean:98]: I guess I would either just leave it in the git history, add a comment why you want to leave this, or make this code that is disabled on a flag and set the flag to off (and add a comment there).

**Review by tobiasgrosser — APPROVED** [review]

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean]: OK, I removed it. Just please don't squash this PR when merging so that the git history is not lost.

Comment (tobiasgrosser) [reply SSA/Core/MLIRSyntax/Parser.lean]: Do you not feel the history in this PR is sufficient? I feel this will make it significantly harder to actually see the real change in the history, will add commits-that-do-not-compile to the history, and also will randomly change our base std library back and forth. I am curious why the history in the PR is not sufficient for someone who wants to investigate this change?

Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean]: because if we squash this PR then we lose this history. It will be only on github and on some random stale branch that's not in the project's history. So if we're working on a file later and want to figure out what happened, why something was there or look at an older version of it, we cannot do 'git log path/to/file.lean`, we have to look it up in every single such PR branch, and that makes it completely unviable in my opinion. 

In my personal opinion `git log` and `git blame` etc are significantly more valuable than `git bisect`, and accurate, fine-grained helps more than a curated history where CI passes for every commit. I could see why for a very large project written in a chaotic language like C++ these priorities would be different, but for an academic project with a much lower PR/commit volume, in a strong typesystem, I think the value of the history is higher (I've never needed to bisect anything in this project, have you?)

Comment (tobiasgrosser) [reply SSA/Core/MLIRSyntax/Parser.lean]: I feel that `git blame` is a lot more efficient as it points to the single PR that had a functional change to the repo rather than hitting some of these 42 commits and having to figure it out. In any case, this seems to need further discussion. Happy to have these eventually and not block this PR over our different perspectives here.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:116]:
```suggestion
        | .none => ".lake"
        | .some buildPath => match buildPath.parent with
```

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:112]: This feels very fragile to me; Can we generalize this, to walk up the parent tree till it finds a `.lake` folder, throwing an error if it doesn't find one, rather than assuming the depth?

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:100]:
```suggestion
  let fileStr := "\n".intercalate lines.toList
```

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:123]: This too feels fragile. Any chance we can as Mac if it's possible to use `Lake` as a library to achieve this?

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:84]:
```suggestion
  match syn with
    | .error msg => throw msg
    | .ok syn => ntparser syn
```
  match Parser.runParserCategory env syntaxcat s with
    | .error msg => throw msg
    | .ok syn => ntparser syn

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:76]:
```suggestion
def elabOp (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Op 0) :=
  unsafe elabIntoEIO (α := MLIR.AST.Op 0) env op0Expr stx
```

File: .github/workflows/ci.yml
Comment (bollu) [reply .github/workflows/ci.yml:42]: @tobiasgrosser for things like this, it's often easier for everyone if you press  <kbd>⌘</kbd> + <kbd> g </kbd> when writing the review, so you can write down a suggested change, as I've done here: 
```suggestion
      - name: Compile `mlirnatural` Executable 🧐
```

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:100]: lol, I don' t know why I left that TODO instead of doing it :thinking:

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:110]: @bollu agreed! However, I' d feel that suggestion is also too fragile, because it will also not work if we at some point have a ` make install`. For that reason I' m not happy with that change, I'd request that we either do it right right (i.e. by having a configurable prefix or an $SSA_HOME or something) or we leave it like it is until we have a proper solution.

**Comment by bollu** [comment]
@goens wondering if you got the chance to follow up with mac.

**Comment by goens** [comment]
@bollu no sorry, I have not, haven't had the time.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [reply SSA/Core/MLIRSyntax/Parser.lean:110]: @goens how about making the user give us the root repo directory, and we navigate relative to this?

**Comment by bollu** [comment]
@goens no worries, was wondering.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:110]: with a default argument? yeah that sounds like a good compromise!

Comment (bollu) [reply SSA/Core/MLIRSyntax/Parser.lean:110]: > with a default argument? yeah that sounds like a good compromise!

Yes, where the default argument, IMO, recursively walks the directory tree to find a `.lake`.

Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:110]: I'm trying to avoid writing code that recusively walks down the directory tree to find a `.lake` because it feels like not so useful work (if the directory structure changes it breaks), I can find almost no use-case for it over what we just have + being able to provide that directory explicitly

**Comment by goens** [comment]
@bollu thanks for the change, that does seem more sane and solves both issues :) so we've got green light to merge, right?

**Comment by bollu** [comment]
Yes, with maybe a final quick skim from @tobiasgrosser

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:4]: @goens can we have a docstring that says what this file does?

File: SSA/Core/MLIRSyntax/GenericParser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/GenericParser.lean:286]:
```suggestion
```
drop dead code.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:132]: My feeling is that the contents of this file as well can be moved into `SSA/Projects/InstCombine/LLVM/Opt.lean`, and can be judiciously refactored out when necessary.

File: SSA/Projects/InstCombine/LLVM/Parser.lean
Comment (goens) [reply SSA/Projects/InstCombine/LLVM/Parser.lean:15]: I have to push back here @bollu. This is conceptually a library, that's why I split it here. The fact that it is *currently* only used by `opt` is just because we haven't implemented anything else, but this could be reused by anything that would want to parse from file.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:132]: Hmm, I disagree. This is not about the verified `opt` executable, this is a library about parsing. I wouldn't mind it if this had been done all in the first place and could be refactored later, but refactoring so that we can later refactor again seems silly :)

File: .github/workflows/ci.yml
Comment (tobiasgrosser) [inline .github/workflows/ci.yml:42]: Needs a more descriptive name than MLIR-natural.

File: .github/workflows/ci.yml
Comment (tobiasgrosser) [inline .github/workflows/ci.yml:46]: Independent cleanup (best to be pushed separately)

File: .github/workflows/ci.yml
Comment (tobiasgrosser) [inline .github/workflows/ci.yml:50]:
```suggestion
          (cd test/bruteforce-correctness && ./run.sh)
```

File: SSA.lean
Comment (tobiasgrosser) [inline SSA.lean:7]: This and the next two changes seem a general refactor that is worth committing separately, no?

File: SSA/Core/MLIRSyntax/GenericParser.lean
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/GenericParser.lean:463]: Is this needed for this PR or can we push this separately (to demonstrate that this is indeed an independent change).

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/Parser.lean:12]: If we link to a zulip thread, can we also commit a 1-2 sentence summary of the thread to the repo, such that we know why it is linked.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/Parser.lean:55]: Newline?

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax/Parser.lean:90]: Should this be resolved, still?

File: .github/workflows/ci.yml
Comment (goens) [reply .github/workflows/ci.yml:42]: That's an unrelated change (renaming `mlirnatural`), since the executable already has this name and has had it for a while. This is change is just disambiguating the two executables in CI

File: SSA/Core/MLIRSyntax/GenericParser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/GenericParser.lean:463]: yes, it's needed for this PR

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:90]: are you asking if this TODO is still necessary or if we can address it and then remove it after having addressed it? I don't know how to do the latter, but if you want to take a stab at it, that'd be very helpful!

File: SSA.lean
Comment (goens) [reply SSA.lean:7]: I feel like this relates to the PR vs commit thing. Absolutely, it should be a different commit, but if PRs and commits are separate things then it's a non-issue.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (goens) [reply SSA/Core/MLIRSyntax/Parser.lean:57]: I don't quite understand this comment. Do you want a newline between the two definitions, or a newline at the end of the file? or do you want something else?

More generally, it sounds like a sort of style question, but we don't have an agreed-upon, fix code style. I don't particularly care for this, happy to commit a suggestion (or you can push a stylistic change that you want to)

Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean:57]:
```suggestion

def op0Expr := (Expr.app (Expr.const `MLIR.AST.Op []) (Expr.const `Nat.zero []))
```

add newline.

Comment (bollu) [reply SSA/Core/MLIRSyntax/Parser.lean:58]: resolved, we agree on mathlib style that uses a newline.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [reply SSA/Core/MLIRSyntax/Parser.lean:93]: @goens I'm not sure what the comment means --- I'm trying to address the TODO, but it's unclear to me what I would need to change to test this. Can I have a sentence or two explaining what environment copy I can try and remove?

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (bollu) [inline SSA/Core/MLIRSyntax/Parser.lean]:
```suggestion
/- see https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Calling.20unification.20from.20meta.20to.20remove.20metavariables

The "metavariable" concept consists of a couple things:
1. the `Expr.mvar` node in an expression that contains a metavariable ID
2. the table of metavariable declarations in the monad, mapping IDs to `Option Expr` (the assignments)

"Assignment" is the process of adding entries to the table, and "instantiation" is the process of replacing `Expr.mvar`s with their assigned values in a given expression.

So we first call `isDefEq` to force unification of metavars, we say `Elab.Term.synthesizeSyntheticMVarsNoPostponing` to also force lean to infer all postponed inferences, and we finally replace the old `expr` with the new instantiated expr with `instantiateMVars`.  

-/
```

File: .github/workflows/ci.yml
Comment (bollu) [reply .github/workflows/ci.yml:42]: @tobiasgrosser I'm going to resolve this, since coming up with good names for these tools is beyond the scope of this PR.

File: lake-manifest.json
Comment (tobiasgrosser) [reply lake-manifest.json]: This seems to have been resolved.

File: .github/workflows/ci.yml
Comment (tobiasgrosser) [reply .github/workflows/ci.yml:42]: I guess this is a strict improvement.

File: SSA/Core/MLIRSyntax/Parser.lean
Comment (tobiasgrosser) [reply SSA/Core/MLIRSyntax/Parser.lean:93]: Just an actionable comment would be enough. Just a bit of information that would people like @bollu enable to pick this issue in the future (or now) would be helpful.

File: SSA.lean
Comment (tobiasgrosser) [reply SSA.lean:7]: @bollu kindly factored this out.

File: .github/workflows/ci.yml
Comment (tobiasgrosser) [reply .github/workflows/ci.yml:46]: @bollu kindly factored this out.

**Comment by bollu** [comment]
@tobiasgrosser this PR LGTM, I've addressed all comments, save for the TODO, which I'm happy to fixup in another PR.

File: SSA.lean
Comment (tobiasgrosser) [inline SSA.lean:9]: An unrelated change. This can be dropped.

File: SSA/Core/MLIRSyntax.lean
Comment (tobiasgrosser) [inline SSA/Core/MLIRSyntax.lean:8]: An unrelated change, this removal of a new line, is not needed here.

**Review by tobiasgrosser — APPROVED** [review]
Nice. I just had some minor comments. It might be useful to manually squash this commit to get a good commit message (otherwise it will just be the titles of all 50 commits).

Thank you @goens @bollu! This is amazing.

File: SSA.lean
Comment (bollu) [reply SSA.lean:9]: @tobiasgrosser I urge you to use the suggestion button :D

**Pull request merged**
